src/DCON/Ode.jl

```diff
@@ -0,0 +1,722 @@
+function ode_sing_init()
+    return
+end
+#     # Declare and initialize local variables
+#     star = fill(' ', mpert)
+#     # local variables
+#     ua = Array{Complex{r8}}(undef, mpert, 2*mpert, 2)
+#     dpsi = 0.0
+#     new = true
+
+#     ising = sing_start
+#     dpsi = singfac_min/abs(nn*sing[ising].q1)*10
+#     odet.psifac = sing[ising].psifac + dpsi
+#     odet.q = sing[ising].q + dpsi*sing[ising].q1
+
+#     # Allocate and initialize solution arrays
+#     u      = zeros(Complex{r8}, mpert, mpert, 2)
+#     du     = zeros(Complex{r8}, mpert, mpert, 2)
+#     u_save = zeros(Complex{r8}, mpert, mpert, 2)
+#     unorm0 = zeros(r8, 2*mpert)
+#     unorm  = zeros(r8, 2*mpert)
+#     index  = zeros(Int, 2*mpert)
+
+#     if old_init
+#         u .= 0.0
+#         for ipert ∈ 1:mpert
+#             u[ipert, ipert, 2] = 1.0
+#         end
+#     else
+#         sing_get_ua(ising, odet.psifac, ua)
+#         # Big slice: u = ua[:, mpert+1:2*mpert,:]
+#         for i = 1:mpert, j = 1:mpert, k = 1:2
+#             u[i, j, k] = ua[i, mpert+j, k]
+#         end
+#     end
+#     u_save .= u
+#     psi_save = psifac
+#     msol = mpert
+#     neq = 4 * mpert * msol
+
+#     # Compute conditions at next singular surface
+#     while true
+#         ising += 1
+#         if ising > msing || psilim < sing[ising ≤ msing ? ising : msing].psifac
+#             break
+#         end
+#         odet.q = sing[ising].q
+#         if mlow ≤ nn*q && mhigh ≥ nn*q
+#             break
+#         end
+#     end
+# This needs to be fixed up
+#     if ising > msing || psilim < sing[ising ≤ msing ? ising : msing].psifac
+#         m1 = round(Int, ctrl.nn*intr.qlim) + round(Int, sign(one, ctrl.nn*equil.sq.fs1[mpsi, 4]))
+#         psimax = psilim * (1-eps)
+#         next_ = "finish"
+#     else
+#         m1 = round(Int, nn*sing[ising].q)
+#         psimax = sing[ising].psifac - singfac_min/abs(nn*sing[ising].q1)
+#         next_ = "cross"
+#     end
+
+#     # Set up integrator parameters
+#     istep = 0
+#     istate = 1
+#     iopt = 1
+#     itask = 5
+#     itol = 2
+#     mf = 10
+#     istate = 1
+
+#     # Set up work arrays
+#     liw = length(iwork)    # iwork must exist!
+#     lrw = 22 + 64*mpert*msol
+#     rwork = zeros(r8, lrw)
+#     atol  = zeros(r8, mpert, msol, 2)
+#     fixfac = zeros(r8, msol, msol)
+#     iwork = zeros(Int, liw)  # or readjust as needed
+#     rwork[1] = psimax
+#     rwork[5] = dpsi
+#     rwork[11] = rwork[5]
+
+#     # Terminate, or in Julia just return (no need for RETURN)
+#     return nothing  # or could return a struct with all these values, for a more Julian approach
+# end
+
+function ode_ideal_cross(ising::Int, odet::OdeState, equil::Equilibrium.PlasmaEquilibrium, intr::DconInternal, ctrl::DconControl)
+    # ...existing code...
+
+    # Fixup solution at singular surface
+    if ctrl.verbose
+        println("psi = $(intr.sing[ising].psifac), q = $(intr.sing[ising].q)")
+    end
+    ode_unorm(true)    
```

Comment:
No inputs and no return value, so what is this doing? If it's operating on mutable global variables, you'll want to rewrite this.